lib/sda/service.go

```diff
@@ -0,0 +1,229 @@
+// it can finds / write everything it needs
+type NewServiceFunc func(c Context, path string) Service
+
+// GenericConfig is a config that can withhold any type of specific configs for
+// protocols. It is passed down to the service NewProtocol function.
+type GenericConfig struct {
+	Type uuid.UUID
+	//Data network.ProtocolMessage
+}
+
+var GenericConfigID = network.RegisterMessageType(GenericConfig{})
+
+// A serviceFactory is used to register a NewServiceFunc
+type serviceFactory struct {
+	cons map[ServiceID]NewServiceFunc
+	// translations between name of a Service and its ServiceID. Used to register a
+	// Service using a name.
+	translations map[string]ServiceID
+	// Inverse mapping of ServiceId => string
+	inverseTr map[ServiceID]string
+}
+
+// ServiceFactory is the global service factory to instantiate Services
+var ServiceFactory = serviceFactory{
+	cons:         make(map[ServiceID]NewServiceFunc),
+	translations: make(map[string]ServiceID),
+	inverseTr:    make(map[ServiceID]string),
+}
+
+// RegisterByName takes an name, creates a ServiceID out of it and store the
+// mapping and the creation function.
+func (s *serviceFactory) Register(name string, fn NewServiceFunc) {
+	id := ServiceID(uuid.NewV5(uuid.NamespaceURL, name))
+	if _, ok := s.cons[id]; ok {
+		// called at init time so better panic than to continue
+		dbg.Lvl1("RegisterService():", name)
+	}
+	s.cons[id] = fn
+	s.translations[name] = id
+	s.inverseTr[id] = name
+}
+
+// RegisterNewService is a wrapper around service factory
+func RegisterNewService(name string, fn NewServiceFunc) {
+	ServiceFactory.Register(name, fn)
+}
+
+// RegisteredServices returns all the services registered
+func (s *serviceFactory) registeredServicesID() []ServiceID {
+	var ids = make([]ServiceID, 0, len(s.cons))
+	for id := range s.cons {
+		ids = append(ids, id)
+	}
+	return ids
+}
+
+// RegisteredServicesByName returns all the names of the services registered
+func (s *serviceFactory) RegisteredServicesName() []string {
+	var names = make([]string, 0, len(s.translations))
+	for n := range s.translations {
+		names = append(names, n)
+	}
+	return names
+}
+
+// ServiceID returns the ServiceID out of the name of the service
+func (s *serviceFactory) ServiceID(name string) ServiceID {
+	var id ServiceID
+	var ok bool
+	if id, ok = s.translations[name]; !ok {
+		return NilServiceID
+	}
+	return id
+}
+
+// Name returns the Name out of the ID
+func (s *serviceFactory) Name(id ServiceID) string {
+	var name string
+	var ok bool
+	if name, ok = s.inverseTr[id]; !ok {
+		return ""
+	}
+	return name
+}
+
+// start launches a new service
+func (s *serviceFactory) start(name string, c Context, path string) (Service, error) {
+	var id ServiceID
+	var ok bool
+	if id, ok = s.translations[name]; !ok {
+		return nil, errors.New("No Service for this name: " + name)
+	}
+	var fn NewServiceFunc
+	if fn, ok = s.cons[id]; !ok {
+		return nil, errors.New("No Service for this id:" + fmt.Sprintf("%v", id))
```

Comment:
Later (not important): Use fmt.Sprintf for the whole string (instead of concatenating with +)